Services/src/CSClassroom/CSClassroom.Service/Projects/SubmissionService.cs

```diff
@@ -230,57 +232,105 @@ public class SubmissionService : ISubmissionService
 		}
 
 		/// <summary>
-		/// Downloads all submissions for a given checkpoint in a section,
-		/// in the form of a zip archive. The archive will also include
-		/// the latest state of repositories for which there is no submission. 
+		/// Fetches list of sections and students with commits that the user
+		/// might want to download.  These can be displayed to the user as lists
+		/// with checkboxes to select from.
+		/// </summary>
+		public async Task<IList<CheckpointDownloadCandidateResult>> GetCheckpointDownloadCandidateListAsync(
+			string classroomName,
+			string projectName,
+			string checkpointName)
+		{
+			// A few round trips to db to get section memberships and submissions...
+
+			var sectionMemberships = await LoadSectionMembershipsAsync(classroomName);
+
+			// Similar to LoadCheckpointAsync, but avoids unnecessary round trips by
+			// not bringing back extra entities.  We only care about the checkpoint id
+			var checkpoint = await _dbContext.Checkpoints
+				.Where(c => c.Project.Classroom.Name == classroomName)
+				.Where(c => c.Project.Name == projectName)
+				.SingleAsync(c => c.Name == checkpointName);
+
+			var submissions = await _dbContext.Submissions
+				.Where
+				(
+					sub => sub.Checkpoint.Id == checkpoint.Id
+				)
+				.Include(sub => sub.Commit)
+				.ToListAsync();
```

Comment:
Done, I think.  I used submissions.ContainsKey(sm.ClassroomMembership.User.Id) instead of submissions[sm.ClassroomMembership.User.Id] to replace the call to Any(), since I just need to know for each user if there is ANY submission for the checkpoint.  I believe this is what you intended.  In any case, GREAT catch for speeding up the query